File: sdk/python/sawtooth_sdk/processor/context.py
# Copyright 2016 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------
from sawtooth_sdk.protobuf.validator_pb2 import Message
from sawtooth_sdk.protobuf import state_context_pb2
from sawtooth_sdk.protobuf import events_pb2
from sawtooth_sdk.processor.exceptions import InternalError
from sawtooth_sdk.processor.exceptions import AuthorizationException


class Context:
    """
    Context provides an interface for getting, setting, and deleting
    validator state. All validator interactions by a handler should be
    through a Context instance.

    Attributes:
        _stream (sawtooth.client.stream.Stream): client grpc communication
        _context_id (str): the context_id passed in from the validator

    """

    def __init__(self, stream, context_id):
        self._stream = stream
        self._context_id = context_id

    def get_state(self, addresses, timeout=None):
        """
        get_state queries the validator state for data at each of the
        addresses in the given list. The addresses that have been set
        are returned in a list.

        Args:
            addresses (list): the addresses to fetch
            timeout: optional timeout, in seconds
        Returns:
            results (list): a list of Entries (address, data), for the
            addresses that have a value

        Raises:
            AuthorizationException
        """
        request = state_context_pb2.TpStateGetRequest(
            context_id=self._context_id,
            addresses=addresses)
        response_string = self._stream.send(
            Message.TP_STATE_GET_REQUEST,
            request.SerializeToString()).result(timeout).content
        response = state_context_pb2.TpStateGetResponse()
        response.ParseFromString(response_string)
        if response.status == \
                state_context_pb2.TpStateGetResponse.AUTHORIZATION_ERROR:
            raise AuthorizationException(
                'Tried to get unauthorized address: {}'.format(addresses))
        entries = response.entries if response is not None else []
        results = [e for e in entries if len(e.data) != 0]
        return results

    def set_state(self, entries, timeout=None):
        """
        set_state requests that each address in the provided dictionary be
        set in validator state to its corresponding value. A list is
        returned containing the successfully set addresses.

        Args:
            entries (dict): dictionary where addresses are the keys and data is
                the value.
            timeout: optional timeout, in seconds

        Returns:
            addresses (list): a list of addresses that were set

        Raises:
            AuthorizationException
        """
        state_entries = [
            state_context_pb2.TpStateEntry(address=e, data=entries[e])
            for e in entries
        ]
        request = state_context_pb2.TpStateSetRequest(
            entries=state_entries,
            context_id=self._context_id).SerializeToString()
        response = state_context_pb2.TpStateSetResponse()
        response.ParseFromString(
            self._stream.send(Message.TP_STATE_SET_REQUEST,
                              request).result(timeout).content)
        if response.status == \
                state_context_pb2.TpStateSetResponse.AUTHORIZATION_ERROR:
            addresses = [e.address for e in state_entries]
            raise AuthorizationException(
                'Tried to set unauthorized address: {}'.format(addresses))
        return response.addresses

    def delete_state(self, addresses, timeout=None):
        """
        delete_state requests that each of the provided addresses be unset
        in validator state. A list of successfully deleted addresses
        is returned.

        Args:
            addresses (list): list of addresses to delete
            timeout: optional timeout, in seconds

        Returns:
            addresses (list): a list of addresses that were deleted

        Raises:
            AuthorizationException
        """
        request = state_context_pb2.TpStateDeleteRequest(
            context_id=self._context_id,
            addresses=addresses).SerializeToString()
        response = state_context_pb2.TpStateDeleteResponse()
        response.ParseFromString(
            self._stream.send(Message.TP_STATE_DELETE_REQUEST,
                              request).result(timeout).content)
        if response.status == \
                state_context_pb2.TpStateDeleteResponse.AUTHORIZATION_ERROR:
            raise AuthorizationException(
                'Tried to delete unauthorized address: {}'.format(addresses))
        return response.addresses

    def add_receipt_data(self, data, timeout=None):
        """Add a blob to the execution result for this transaction.

        Args:
            data (bytes): The data to add.
        """
        request = state_context_pb2.TpReceiptAddDataRequest(
            context_id=self._context_id,
            data=data).SerializeToString()
        response = state_context_pb2.TpReceiptAddDataResponse()
        response.ParseFromString(
            self._stream.send(
                Message.TP_RECEIPT_ADD_DATA_REQUEST,
                request).result(timeout).content)
        if response.status == state_context_pb2.TpReceiptAddDataResponse.ERROR:
            raise InternalError(
                "Failed to add receipt data: {}".format((data)))

    def add_event(self, event_type, attributes=None, data=None, timeout=None):
        """Add a new event to the execution result for this transaction.

        Args:
            event_type (str): This is used to subscribe to events. It should be
                globally unique and describe what, in general, has occured.
            attributes (list of (str, str) tuples): Additional information
                about the event that is transparent to the validator.
                Attributes can be used by subscribers to filter the type of
                events they receive.
            data (bytes): Additional information about the event that is opaque
                to the validator.
        """
        if attributes is None:
            attributes = []

        event = events_pb2.Event(
            event_type=event_type,
            attributes=[
                events_pb2.Event.Attribute(key=key, value=value)
                for key, value in attributes
            ],
            data=data,
        )
        request = state_context_pb2.TpEventAddRequest(
            context_id=self._context_id, event=event).SerializeToString()
        response = state_context_pb2.TpEventAddResponse()
        response.ParseFromString(
            self._stream.send(
                Message.TP_EVENT_ADD_REQUEST,
                request).result(timeout).content)
        if response.status == state_context_pb2.TpEventAddResponse.ERROR:
            raise InternalError(
                "Failed to add event: ({}, {}, {})".format(
                    event_type, attributes, data))

File: sdk/go/src/sawtooth_sdk/processor/context.go
/**
 * Copyright 2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ------------------------------------------------------------------------------
 */

package processor

import (
	"fmt"
	"github.com/golang/protobuf/proto"
	"sawtooth_sdk/messaging"
	"sawtooth_sdk/protobuf/events_pb2"
	"sawtooth_sdk/protobuf/state_context_pb2"
	"sawtooth_sdk/protobuf/validator_pb2"
)

// Context provides an abstract interface for getting and setting validator
// state. All validator interactions by a handler should be through a Context
// instance. Currently, the Context class is NOT thread-safe and Context classes
// may not share the same messaging.Connection object.
type Context struct {
	connection messaging.Connection
	contextId  string
}

type Attribute struct {
	Key   string
	Value string
}

// Construct a new context object given an initialized Stream and Context ID.
func NewContext(connection messaging.Connection, contextId string) *Context {
	return &Context{
		connection: connection,
		contextId:  contextId,
	}
}

// GetState queries the validator state for data at each of the addresses in the
// given slice. A string->[]byte map is returned. If an address is not set,
// it will not exist in the map.
//
//     results, err := context.Get(addresses)
//     if err != nil {
//         fmt.Println("Error getting data!")
//     }
//     data, ok := results[address]
//     if !ok {
//         fmt.Prinln("No data stored at address!")
//     }
//
func (self *Context) GetState(addresses []string) (map[string][]byte, error) {
	// Construct the message
	request := &state_context_pb2.TpStateGetRequest{
		ContextId: self.contextId,
		Addresses: addresses,
	}
	bytes, err := proto.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("Failed to marshal TpStateGetRequest: %v", err)
	}

	// Send the message and get the response
	corrId, err := self.connection.SendNewMsg(
		validator_pb2.Message_TP_STATE_GET_REQUEST, bytes,
	)
	if err != nil {
		return nil, fmt.Errorf("Failed to send TpStateGetRequest: %v", err)
	}

	_, msg, err := self.connection.RecvMsgWithId(corrId)
	if msg.GetCorrelationId() != corrId {
		return nil, fmt.Errorf(
			"Expected message with correlation id %v but got %v",
			corrId, msg.GetCorrelationId(),
		)
	}

	if msg.GetMessageType() != validator_pb2.Message_TP_STATE_GET_RESPONSE {
		return nil, fmt.Errorf(
			"Expected TpStateGetResponse but got %v", msg.GetMessageType(),
		)
	}

	// Parse the result
	response := &state_context_pb2.TpStateGetResponse{}
	err = proto.Unmarshal(msg.GetContent(), response)
	if err != nil {
		return nil, fmt.Errorf("Failed to unmarshal TpStateGetResponse: %v", err)
	}

	// Use a switch in case new Status values are added
	switch response.Status {
	case state_context_pb2.TpStateGetResponse_AUTHORIZATION_ERROR:
		return nil, &AuthorizationException{Msg: fmt.Sprint("Tried to get unauthorized address: ", addresses)}
	}

	// Construct and return a map
	results := make(map[string][]byte)
	for _, entry := range response.GetEntries() {
		if len(entry.GetData()) != 0 {
			results[entry.GetAddress()] = entry.GetData()
		}
	}

	return results, nil
}

// SetState requests that each address in the validator state be set to the given
// value. A slice of addresses set is returned or an error if there was a
// problem setting the addresses. For example:
//
//     responses, err := context.Set(dataMap)
//     if err != nil {
//         fmt.Println("Error setting addresses!")
//     }
//     set, ok := results[address]
//     if !ok {
//         fmt.Prinln("Address was not set!")
//     }
//
func (self *Context) SetState(pairs map[string][]byte) ([]string, error) {
	// Construct the message
	entries := make([]*state_context_pb2.TpStateEntry, 0, len(pairs))
	for address, data := range pairs {
		entries = append(entries, &state_context_pb2.TpStateEntry{
			Address: address,
			Data:    data,
		})
	}

	request := &state_context_pb2.TpStateSetRequest{
		ContextId: self.contextId,
		Entries:   entries,
	}
	bytes, err := proto.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("Failed to marshal: %v", err)
	}

	// Send the message and get the response
	corrId, err := self.connection.SendNewMsg(
		validator_pb2.Message_TP_STATE_SET_REQUEST, bytes,
	)
	if err != nil {
		return nil, fmt.Errorf("Failed to send set: %v", err)
	}

	_, msg, err := self.connection.RecvMsgWithId(corrId)
	if msg.GetCorrelationId() != corrId {
		return nil, fmt.Errorf(
			"Expected message with correlation id %v but got %v",
			corrId, msg.GetCorrelationId(),
		)
	}

	if msg.GetMessageType() != validator_pb2.Message_TP_STATE_SET_RESPONSE {
		return nil, fmt.Errorf(
			"Expected TP_STATE_SET_RESPONSE but got %v", msg.GetMessageType(),
		)
	}

	// Parse the result
	response := &state_context_pb2.TpStateSetResponse{}
	err = proto.Unmarshal(msg.Content, response)
	if err != nil {
		return nil, fmt.Errorf("Failed to unmarshal TpStateSetResponse: %v", err)
	}

	// Use a switch in case new Status values are added
	switch response.GetStatus() {
	case state_context_pb2.TpStateSetResponse_AUTHORIZATION_ERROR:
		addresses := make([]string, 0, len(pairs))
		for a, _ := range pairs {
			addresses = append(addresses, a)
		}
		return nil, &AuthorizationException{Msg: fmt.Sprint("Tried to set unauthorized address: ", addresses)}
	}

	return response.GetAddresses(), nil
}

// DeleteState requests that each address in the validator state be
// deleted from state. A slice of addresses deleted is returned or an
// error if there was a problem deleting the addresses. For example:
//
//     responses, err := context.DeleteState(dataMap)
//     if err != nil {
//         fmt.Println("Error deleting addresses!")
//     }
//     delete, ok := results[address]
//     if !ok {
//         fmt.Prinln("Address was not deleted!")
//     }
//
func (self *Context) DeleteState(addresses []string) ([]string, error) {
	// Construct the message
	request := &state_context_pb2.TpStateDeleteRequest{
		ContextId: self.contextId,
		Addresses: addresses,
	}

	bytes, err := proto.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("Failed to marshal TpStateDeleteRequest: %v", err)
	}

	// Send the message and get the response
	corrId, err := self.connection.SendNewMsg(
		validator_pb2.Message_TP_STATE_DELETE_REQUEST, bytes,
	)
	if err != nil {
		return nil, fmt.Errorf("Failed to send TpStateDeleteRequest: %v", err)
	}

	_, msg, err := self.connection.RecvMsgWithId(corrId)
	if msg.GetCorrelationId() != corrId {
		return nil, fmt.Errorf(
			"Expected message with correlation id %v but got %v",
			corrId, msg.GetCorrelationId(),
		)
	}

	if msg.GetMessageType() != validator_pb2.Message_TP_STATE_DELETE_RESPONSE {
		return nil, fmt.Errorf(
			"Expected TpStateDeleteResponse but got %v", msg.GetMessageType(),
		)
	}

	// Parse the result
	response := &state_context_pb2.TpStateDeleteResponse{}
	err = proto.Unmarshal(msg.GetContent(), response)
	if err != nil {
		return nil, fmt.Errorf("Failed to unmarshal TpStatedDeleteResponse: %v", err)
	}

	// Use a switch in case new Status values are added
	switch response.Status {
	case state_context_pb2.TpStateDeleteResponse_AUTHORIZATION_ERROR:
		return nil, &AuthorizationException{Msg: fmt.Sprint("Tried to get unauthorized address: ", addresses)}
	}

	return response.GetAddresses(), nil
}

// Add a blob to the execution result for this transaction.
func (self *Context) AddReceiptData(data []byte) error {
	// Append the data to the transaction receipt and set the type
	request := &state_context_pb2.TpReceiptAddDataRequest{
		ContextId: self.contextId,
		Data:      data,
	}
	bytes, err := proto.Marshal(request)
	if err != nil {
		return fmt.Errorf("Failed to marshal: %v", err)
	}

	// Send the message and get the response
	corrId, err := self.connection.SendNewMsg(
		validator_pb2.Message_TP_RECEIPT_ADD_DATA_REQUEST, bytes,
	)
	if err != nil {
		return fmt.Errorf("Failed to add receipt data: %v", err)
	}

	_, msg, err := self.connection.RecvMsgWithId(corrId)
	if msg.GetCorrelationId() != corrId {
		return fmt.Errorf(
			"Expected message with correlation id %v but got %v",
			corrId, msg.GetCorrelationId(),
		)
	}

	if msg.GetMessageType() != validator_pb2.Message_TP_RECEIPT_ADD_DATA_RESPONSE {
		return fmt.Errorf(
			"Expected TP_RECEIPT_ADD_DATA_RESPONSE but got %v", msg.GetMessageType(),
		)
	}

	// Parse the result
	response := &state_context_pb2.TpReceiptAddDataResponse{}
	err = proto.Unmarshal(msg.Content, response)
	if err != nil {
		return fmt.Errorf("Failed to unmarshal TpReceiptAddDataResponse: %v", err)
	}

	// Use a switch in case new Status values are added
	switch response.GetStatus() {
	case state_context_pb2.TpReceiptAddDataResponse_ERROR:
		return fmt.Errorf("Failed to add receipt data")
	}

	return nil
}

// Add a new event to the execution result for this transaction.
func (self *Context) AddEvent(event_type string, attributes []Attribute, event_data []byte) error {
	event_attributes := make([]*events_pb2.Event_Attribute, 0, len(attributes))
	for _, attribute := range attributes {
		event_attributes = append(
			event_attributes,
			&events_pb2.Event_Attribute{
				Key:   attribute.Key,
				Value: attribute.Value,
			},
		)
	}

	event := &events_pb2.Event{
		EventType:  event_type,
		Attributes: event_attributes,
		Data:       event_data,
	}

	// Construct message
	request := &state_context_pb2.TpEventAddRequest{
		ContextId: self.contextId,
		Event:     event,
	}
	bytes, err := proto.Marshal(request)
	if err != nil {
		return fmt.Errorf("Failed to marshal: %v", err)
	}

	// Send the message and get the response
	corrId, err := self.connection.SendNewMsg(
		validator_pb2.Message_TP_EVENT_ADD_REQUEST, bytes,
	)
	if err != nil {
		return fmt.Errorf("Failed to add event: %v", err)
	}

	_, msg, err := self.connection.RecvMsgWithId(corrId)
	if msg.GetCorrelationId() != corrId {
		return fmt.Errorf(
			"Expected message with correlation id %v but got %v",
			corrId, msg.GetCorrelationId(),
		)
	}

	if msg.GetMessageType() != validator_pb2.Message_TP_EVENT_ADD_RESPONSE {
		return fmt.Errorf(
			"Expected TP_EVENT_ADD_RESPONSE but got %v", msg.GetMessageType(),
		)
	}

	// Parse the result
	response := &state_context_pb2.TpEventAddResponse{}
	err = proto.Unmarshal(msg.Content, response)
	if err != nil {
		return fmt.Errorf("Failed to unmarshal TpEventAddResponse: %v", err)
	}

	// Use a switch in case new Status values are added
	switch response.GetStatus() {
	case state_context_pb2.TpEventAddResponse_ERROR:
		return fmt.Errorf("Failed to add event")
	}

	return nil
}
